### PR TITLE
Fixed norm

### LINKED ISSUE #593: incorrect norms for vectors with typemin(Int)
Our L1 and Linf norms give a negative result for `typemin(Int)` (since the `abs` function wraps around), even though they return a floating-point result (see JuliaLang/LinearAlgebra.jl#575) that could represent the correct answer. 
```
julia> norm([typemin(Int), typemin(Int)], Inf) # wrong
-9.223372036854776e18

julia> norm([typemin(Int), typemin(Int)], 1) # wrong
-1.8446744073709552e19

julia> norm([typemin(Int), typemin(Int)], 2) # correct
1.3043817825332783e19
```
Suggested solution: convert to `float` *before* computing `abs` in these functions.

The `hypot` function [has a similar problem](https://github.com/JuliaLang/julia/pull/30301#issuecomment-449059852).

cc @andreasnoack

### LINKED ISSUE #30466: incorrect norms for vectors with typemin(Int)
Our L1 and Linf norms give a negative result for `typemin(Int)` (since the `abs` function wraps around), even though they return a floating-point result (see #29685) that could represent the correct answer. 
```
julia> norm([typemin(Int), typemin(Int)], Inf) # wrong
-9.223372036854776e18

julia> norm([typemin(Int), typemin(Int)], 1) # wrong
-1.8446744073709552e19

julia> norm([typemin(Int), typemin(Int)], 2) # correct
1.3043817825332783e19
```
Suggested solution: convert to `float` *before* computing `abs` in these functions.

The `hypot` function [has a similar problem](https://github.com/JuliaLang/julia/pull/30301#issuecomment-449059852).

cc @andreasnoack

### PR DESCRIPTION
**Pull request by raghav9-97**
Fixes JuliaLang/LinearAlgebra.jl#593.

### CODE REVIEW

**Review by stevengj — COMMENTED** [review]
This isn’t correct, I think: the `float` function isn’t defined for an arbitrary iterators, and in any case might make a copy of a whole array.

We need to call `float` on the individual *elements* inside the loops of the norm functions.

File: stdlib/LinearAlgebra/src/generic.jl [hunk 1]
```diff
@@ -273,40 +273,40 @@ diag(A::AbstractVector) = throw(ArgumentError("use diagm instead of diag to cons
 # special cases of norm; note that they don't need to handle isempty(x)
 function generic_normMinusInf(x)
     (v, s) = iterate(x)::Tuple
-    minabs = norm(v)
+    minabs = norm(float(v))
```
Comment (stevengj) [inline stdlib/LinearAlgebra/src/generic.jl]: this isn't right either, since `v` might not be a scalar.

**Comment by stevengj** [comment]
We might need to call something like `floatnorm(v)`, where we define
```jl
floatnorm(x) = norm(x) # fallback
floatnorm(x::Number) = norm(float(x)) # avoid wraparound for `typemin(Int)` values
```
(There's also the question of what the performance hit for this is, and whether it is worth it.)

**Comment by raghav9-97** [comment]
@stevengj The failing Quaternion array test is due to absence of AbstractFloat(::Quaternion{Float64}) declaration which would again fall under ```floatnorm(x::Number) = norm(float(x))``` and this will again throw the exception.
So according to me, this would not help us.

**Comment by stevengj** [comment]
Note that the quaternion type is also susceptible to wraparound problems for `typemin(Int)`, and arguably it should define a `float` method.

Two questions:

* do we want to require all `Number` types to define a `float` method to fix this problem?

* The `float(x::Number) = AbstractFloat(x)` fallback is arguably wrong for non-`Real` types.   Should we change it to `float(x::Real) = AbstractFloat(x)`?   Maybe the generic fallback can be `float(x::Number) = promote(x, float(real(x)))[1]` or similar?

**Comment by raghav9-97** [comment]
```float(x::Number) = promote(x, float(real(x)))[1]``` also throws an error as there is no method real for Quaternion. Maybe we can generalize float for Real types like ```float(x::Real) = AbstractFloat(x)``` and for Quaternion we can define something like this ```AbstractFloat(x::T) where {T<:Number} = x```.

**Comment by stevengj** [comment]
Arguably, any [serious quaternion type should have a `real` function](https://github.com/JuliaGeometry/Quaternions.jl/blob/14bcda9664e6f38290c2281b22e21391c58e446e/src/Quaternion.jl#L42-L44) that returns the real part.

**Comment by raghav9-97** [comment]
@stevengj What should be the output of ```real(Quaternion{Float64}(0.0, 0.20727281732800007, -0.18047655175799998, 0.006155888157999945)``` according to you as it is obtained after ```(v,s) = iterate(::Array{Quaternion{Float64,1})``` as v and then we need to apply ```norm(float(v))```?

**Comment by raghav9-97** [comment]
@stevengj Maybe we can explicitly define float for Quaternion type just like other function are explicitly defined for it like abs, real or conj.

**Comment by stevengj** [comment]
@raghav9-97, I don't think that's correct.  `AbstractFloat(x)` should presumably return a subtype of `AbstractFloat`, which is why `float` is more appropriate.   It's also why that fallback of `float` should be restricted to `Real` types, as I said above.

Similarly, `AbstractFloat(3+4im)` is not defined but `float(3+4im)` works.

**Comment by andreasnoack** [comment]
Given that we currently promote to floats in `norm` for array and iterator inputs, I think it would be reasonable to make `norm(::Number)` return floats as well by converting the input to float before calling `abs`.

We thereby require that any number type used with `norm` defines `float` but I think many do that already. I know that some autodiff number had to define `float` for other reasons so it is probably okay to require. Hence, I'd just add a `float` definition to out test quaternion type.

A final comment, given our choice of integer arithmetic, I think #30466 is actually only an issue because we convert to floats in `norm`. A negative result would be in line with `abs` returning a negative result but since the result is converted to floats, it's no longer expected that integer arithmetic has been used. Something along the lines of https://github.com/JuliaLang/julia/issues/29685 might, therefore, be a prettier, if not better solution in the future.

**Comment by andreasnoack** [comment]
Yes. That definition for `float(Quaternion)` should be fine. I'm a bit surprised that many tests would need to be changed. Can you point me to some them?

**Comment by raghav9-97** [comment]
I have not yet checked for the tests which needs to be changed maybe there aren't any. In addition to the Quaternion float definition and converting input to float before ```abs``` what else do we need to change?

**Comment by andreasnoack** [comment]
I suspect that you can just remove all the `v = isa(v, Number) ? float(v) : v` lines once you've made `norm` return a float.

**Comment by raghav9-97** [comment]
Thanks for the help @andreasnoack.

File: stdlib/LinearAlgebra/test/generic.jl [hunk 2]
```diff
@@ -225,6 +225,12 @@ end
     end
 end
 
+@testset "Issue #30466" begin
+    @test norm([typemin(Int), typemin(Int)], Inf) == 9.223372036854776e18
+    @test norm([typemin(Int), typemin(Int)], -Inf) == 9.223372036854776e18
```
Comment (andreasnoack) [inline stdlib/LinearAlgebra/test/generic.jl]: Did we discuss this anywhere? I'm not sure we should allow negative `p` arguments so probably better to leave out this test.

**Review by andreasnoack — APPROVED** [review]

**Comment by raghav9-97** [comment]
Is it now good enough to be merged?

**Comment by stevengj** [comment]
Two issues:

* Making `norm(x::Number)` return a float is a breaking change, so I don't think we can do this?

* What is the performance impact on `norm(array, 1)` for an array of `Int`?

**Comment by raghav9-97** [comment]
What's the status of this PR?

**Comment by stevengj** [comment]
tests are failing?

**Comment by stevengj** [comment]
Also, what is the performance impact (if any) for `@btime norm($(rand(Int, 100)), 1)`?

File: stdlib/LinearAlgebra/test/generic.jl [hunk 3]
```diff
@@ -225,6 +225,11 @@ end
     end
 end
 
+@testset "Issue #30466" begin
+    @test norm([typemin(Int), typemin(Int)], Inf) == 9.223372036854776e18
```
Comment (andreasnoack) [inline stdlib/LinearAlgebra/test/generic.jl]: `Int` is `Int32` on 32 bit systems so make the rhs here `float(typemax(Int))` and `2float(typemax(Int))` in the line below.

Comment (stevengj) [reply stdlib/LinearAlgebra/test/generic.jl]: You mean `-float(typemin(Int))`, I guess (since `typemin` and `typemax` are not equal in magnitude).

**Comment by raghav9-97** [comment]
Is this PR now ready to be merged?

**Comment by andreasnoack** [comment]
I believe we are still waiting for the benchmarks requested in 
https://github.com/JuliaLang/julia/pull/30481#issuecomment-451440730

**Comment by raghav9-97** [comment]
Benchmarks for new norm function
```
julia> @btime norm($(rand(Int, 100)), 1)
  95.511 ns (0 allocations: 0 bytes)
4.830041879970702e20

julia> @benchmark norm($(rand(Int, 100)), 1)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     93.957 ns (0.00% GC)
  median time:      94.664 ns (0.00% GC)
  mean time:        100.960 ns (0.00% GC)
  maximum time:     331.780 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     951
```
Benchmarks for old norm function
```
julia> @btime norm($(rand(Int, 100)), 1)
  103.450 ns (0 allocations: 0 bytes)
4.644827316613159e20

julia> @benchmark norm($(rand(Int, 100)), 1)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     103.400 ns (0.00% GC)
  median time:      103.988 ns (0.00% GC)
  mean time:        109.282 ns (0.00% GC)
  maximum time:     468.727 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     927
```

**Review by stevengj — APPROVED** [review]

**Comment by raghav9-97** [comment]
Thanks for helping me out @andreasnoack @stevengj.

**Pull request merged**